### DISCUSSION TITLE
clGetProgramInfo(CL_PROGRAM_BINARIES) builds generic WG kernel, causing extra compilation on already built kernels

### DISCUSSION THREAD

**Discussion by matthiasdiener** [pocl/pocl#1483] [General]
When calling `clGetProgramInfo(CL_PROGRAM_BINARIES)` (e.g., to store built kernels in a cache like PyOpenCL does), a generic WG function is built in addition to the device-specific code:

https://github.com/pocl/pocl/blob/903c83464831db70bb721d13ff2811b4485fe979/lib/CL/devices/common_driver.c#L1017-L1019

This essentially doubles compile time when getting binaries compared to just building and running a kernel, which can be especially noticeable with large kernels, as in the following example code:

```python
import pyopencl as cl
import numpy as np
import pyopencl.array as cl_array
import time

rng = np.random.default_rng()
a = rng.random(5, dtype=np.float32)
b = rng.random(5, dtype=np.float32)

ctx = cl.create_some_context()
queue = cl.CommandQueue(ctx)

a_dev = cl_array.to_device(queue, a)
b_dev = cl_array.to_device(queue, b)
dest_dev = cl_array.empty_like(a_dev)

start = time.time()

# Build & run a large kernel
prg = cl.Program(ctx, """
                 __kernel void f(__global const float *a,
                 __global const float *b, __global float *c) {int gid = get_global_id(0);
                 """ + "c[gid] = a[gid] + b[gid];\n" * 2000 + "}").build()

knl = prg.f
knl(queue, a.shape, None, a_dev.data, b_dev.data, dest_dev.data)

queue.finish()

stop1 = time.time()

print(stop1 - start)

# Get the binaries
prg.get_info(cl.program_info.BINARIES)

stop2 = time.time()
print(stop2 - stop1)
```

```console
$ python t.py
4.9807448387146    # initial compilation + kernel execution
4.343946933746338  # second compilation due to clGetProgramInfo(CL_PROGRAM_BINARIES)
```

I was wondering, is there a way to disable this behavior, or another way to get the binaries without triggering the extra compilation?

cc: @inducer, https://github.com/pocl/pocl/issues/1313

**Comment by franz** [comment]
The root of the issue is that the `clGetProgramInfo` triggers a compilation of "generic" binaries, which are able to work with any workgroup size. It is necessary to have them in the binaries, as in general we can't know which workgroup size will be used with the loaded binaries, and the binaries might be loaded by PoCL which was compiled without LLVM support. OTOH, calling clBuildProgram + clEnqueueNDRange results in compilation of "specialized" binaries which are compiled for one specific workgroup size.

You could try: `export POCL_WORK_GROUP_SPECIALIZATION=0`, this should disable building of specialized binaries. `clGetProgramInfo` should then be instant, at the cost of possibly somewhat slower kernels.

**Reply by pjaaskel** [reply]
It would be interesting to know how much slower the non-local-size-specialized WG-functions are nowadays with the latest LLVM's loopvec. The local-size specialization is kind of a historic artifact from the original VLIW target where we wanted to fully unroll WIs due to only having multi-issue. For loop-vec now LLVM has advanced tricks for specializing the loops which it cannot vectorize due to the runtime defined loop iteration count. So my guess is that the end result (in amount of total code compiled) through loop vectorizer for loops it can vectorize is close to similar what PoCL produces since it might copy the loops inside the WG functions (one with an unknown one and one with known bounds for SIMD width) and does not generate separate WG functions like PoCL does. Something to measure at some point.
